src/store/image-stats.ts

```diff
@@ -0,0 +1,222 @@
+  hist.reduce((acc, val) => {
+    const currentSum = acc + val;
+    cumulativeHist.push(currentSum);
+    return currentSum;
+  }, 0);
+
+  const width = (max - min + 1) / WL_HIST_BINS;
+  const totalCount = scalarData.length;
+
+  return Object.fromEntries(
+    Object.entries(WLAutoRanges).map(([key, percentage]) => {
+      const lowerBound = percentage * 0.01 * totalCount;
+      const upperBound = (1 - percentage * 0.01) * totalCount;
+
+      const startIdx = cumulativeHist.findIndex((v) => v >= lowerBound);
+      const endIdx = cumulativeHist.findIndex((v) => v >= upperBound);
+
+      const start = Math.max(min, min + width * startIdx);
+      const end = Math.min(max, min + width * (endIdx + 1)); // Adjusted end calculation
+      return [key, [start, end] as [number, number]];
+    })
+  );
+}
+
+export const useImageStatsStore = defineStore('image-stats', () => {
+  const stats = reactive<Record<string, ImageStats>>({});
+  const imageCacheStore = useImageCacheStore();
+  const messageStore = useMessageStore();
+
+  const statsEffectScope: Record<string, EffectScope> = {};
+  const autoRangeComputations: Record<
+    string,
+    Promise<Record<string, [number, number]>>
+  > = {};
+
+  const internalSetScalarRange = (
+    imageID: string,
+    min: number,
+    max: number
+  ) => {
+    stats[imageID] = {
+      ...stats[imageID],
+      scalarMin: min,
+      scalarMax: max,
+    };
+  };
+
+  const internalSetAutoRangeValues = (
+    imageID: string,
+    autoValues: Record<string, [number, number]>
+  ) => {
+    stats[imageID] = {
+      ...stats[imageID],
+      autoRangeValues: autoValues,
+    };
+  };
+
+  const internalRemoveStats = (imageID: string) => {
+    delete stats[imageID];
+  };
+
+  const setupImageWatchers = (id: string) => {
+    const { imageData, isLoading: isImageLoading } = useImage(
+      computed(() => id)
+    );
+
+    const activeScalars = computed(() =>
+      imageData.value?.getPointData()?.getScalars()
+    );
+    const scalarRange = vtkFieldRef(activeScalars, 'range');
+
+    watch(
+      scalarRange,
+      (range) => {
+        if (range) {
+          internalSetScalarRange(id, range[0], range[1]);
+        }
+      },
+      { immediate: true }
+    );
+
+    const triggerAutoRangeComputation = (image: vtkImageData) => {
+      autoRangeComputations[id] = computeAutoRangeValues(image);
+
+      autoRangeComputations[id]
+        .then((autoValues) => {
+          if (imageCacheStore.imageIds.includes(id)) {
+            // not deleted yet, save values
+            internalSetAutoRangeValues(id, autoValues);
+          }
+        })
+        .catch((error) => {
+          console.error(
+            `[ImageStatsStore] Auto range computation for image ${id} FAILED:`,
+            error
+          );
+          messageStore.addError(
+            `Auto range computation failed for image ${id}`,
+            error instanceof Error ? error : String(error)
```

Comment:
minor improvement: You can use `ensureError(error)` here

Reply:
fixed